src/accounts.cpp

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2020 YADRO
+
+#include "account_entry.hpp"
+#include "account_list.hpp"
+#include "accounts.hpp"
+
+#include <set>
+#include <string>
+
+namespace fs = std::filesystem;
+
+/** @brief Directory name for accounts files. */
+static const char* accountsDir = "etc";
+/** @brief Name of the group file. */
+static const char* groupFile = "group";
+/** @brief Name of the passwd file. */
+static const char* passwdFile = "passwd";
+/** @brief Name of the shadow file. */
+static const char* shadowFile = "shadow";
+
+/** @brief Minimal value for UID. Most of the Linux systems (and OpenBMC too)
+ *         have defined it to 1000, see UID_MIN from /etc/login.defs.
+ */
+static constexpr uint16_t minUserId = 1000;
+
+// clang-format off
+/** @brief List of users, that created by OpenBMC but can be changed by an end
+ *         user. */
+static const std::set<const char*> allowedUsers = {
+    "admin"
+};
+/** @brief List of groups which membership can be modified by an end user. */
+static const std::set<const char*> allowedGroups = {
+    "priv-admin",
+    "priv-operator",
+    "priv-user",
+    "ipmi",
+    "redfish",
+    "web",
+};
+// clang-format on
+
+using Groups = AccountList<GroupEntry>;
+using Passwd = AccountList<PasswdEntry>;
+using Shadow = AccountList<ShadowEntry>;
+
+Accounts::Accounts(const fs::path& srcRoot, const fs::path& dstRoot,
+                   const fs::path& roRoot) :
+    srcDir(srcRoot / accountsDir),
+    dstDir(dstRoot / accountsDir), roDir(roRoot / accountsDir)
+{
+    fs::create_directories(dstDir);
+}
+
+void Accounts::backup()
+{
+    backupGroup();
+    backupPasswd();
+    backupShadow();
+}
+
+void Accounts::restore()
+{
+    restoreGroup();
+    restorePasswd();
+    restoreShadow();
+}
+
+void Accounts::backupGroup()
+{
+    Groups bk;
+    bk.load(srcDir / groupFile);
+
+    // Remove groups that are not in the white list
+    bk.remove(allowedGroups, false);
+
+    bk.save(dstDir / groupFile);
+}
+
+void Accounts::backupPasswd()
+{
+    Passwd bk;
+    bk.load(srcDir / passwdFile);
+
+    Passwd ro;
+    ro.load(roDir / passwdFile);
+    ro.remove(allowedUsers, true); // Exception for modifiable user accounts
+
+    // Remove build-in accounts
+    bk.remove(ro, true);
+
+    bk.save(dstDir / passwdFile);
+}
+
+void Accounts::backupShadow()
+{
+    Shadow bk;
+    bk.load(srcDir / shadowFile);
+
+    Shadow ro;
+    ro.load(roDir / shadowFile);
+    ro.remove(allowedUsers, true); // Exception for modifiable user accounts
+
+    // Remove build-in accounts
+    bk.remove(ro, true);
+
+    bk.save(dstDir / shadowFile);
+}
+
+void Accounts::restoreGroup()
+{
+    Groups bk;
+    bk.load(srcDir / groupFile);
+
+    Groups rst;
+    rst.load(roDir / groupFile);
+
+    for (const auto& it : allowedGroups)
```

Comment:
>  Why do you save the groups from RO?

Because we need to backup/restore information about memberships.

> Do you try to save the groups which was removed after the BMC install? :)

We don't have to. Groups can not be created or removed by an end user. LDAP is not supported yet.